### PR TITLE
op-service/backoff: Refactor to use time.Duration instead of float/int.

### PR DESCRIPTION
**Pull request by Rcontre360**
changed backoff/strategies.go and backoff/strategies_test.go, only types and a function changed

**Description**

No features added. Just refactored backoff package to allow only types from time.Duration on time related variables. 

**Tests**

Only modified a single test to use time.Duration instead of int.

**Metadata**

- Fixes #[6730](https://github.com/ethereum-optimism/optimism/issues/6730#issuecomment-1678950667)

### CODE REVIEW

File: op-service/backoff/strategies.go [hunk 1]
```diff
@@ -43,8 +43,8 @@ func (e *ExponentialStrategy) Duration(attempt int) time.Duration {
 
 func Exponential() Strategy {
 	return &ExponentialStrategy{
-		Max:       10000,
-		MaxJitter: 250,
+		Max:       time.Duration(10000),
+		MaxJitter: time.Duration(250),
```
Comment (protolambda) [inline op-service/backoff/strategies.go]:
```suggestion
		Max:       time.Duration(10000 * time.Millisecond),
		MaxJitter: time.Duration(250 * time.Millisecond),
```

File: op-service/backoff/strategies.go [hunk 2]
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
 // function is min(e.Min + (2^attempt * 1000) + randBetween(0, e.MaxJitter), e.Max)
 type ExponentialStrategy struct {
 	// Min is the minimum amount of time to wait between attempts in ms.
```
Comment (protolambda) [inline op-service/backoff/strategies.go]:
```suggestion
	// Min is the minimum amount of time to wait between attempts.
```

File: op-service/backoff/strategies.go [hunk 3]
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
...
-	Min float64
+	Min time.Duration
 
 	// Max is the maximum amount of time to wait between attempts in ms.
```
Comment (protolambda) [inline op-service/backoff/strategies.go]:
```suggestion
	// Max is the maximum amount of time to wait between attempts.
```

File: op-service/backoff/strategies.go [hunk 4]
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
...
-	Max float64
+	Max time.Duration
 
 	// MaxJitter is the maximum amount of random jitter to insert between
 	// attempts in ms.
```
Comment (protolambda) [inline op-service/backoff/strategies.go]:
```suggestion
	// MaxJitter is the maximum amount of random jitter to insert between attempts.
```

File: op-service/backoff/strategies.go [hunk 5]
```diff
@@ -17,23 +17,23 @@ type Strategy interface {
...
-	var jitter int
+	var jitter time.Duration
 	if e.MaxJitter > 0 {
-		jitter = rand.Intn(e.MaxJitter)
+		jitter = time.Duration(rand.Int63n(e.MaxJitter.Milliseconds()))
 	}
-	dur := e.Min + (math.Pow(2, float64(attempt)) * 1000)
-	dur += float64(jitter)
+	dur := e.Min + time.Duration((math.Pow(2, float64(attempt)) * 1000))
+	dur += jitter
```
Comment (protolambda) [inline op-service/backoff/strategies.go]: Also needs corrections. You can't just convert from int to time.Duration, it has to scale with the correct unit.

File: op-service/backoff/strategies_test.go [hunk 6]
```diff
@@ -14,8 +14,8 @@ func TestExponential(t *testing.T) {
 		MaxJitter: 0,
 	}
 
-	durations := []int{4, 5, 7, 10, 10}
+	durations := []time.Duration{4, 5, 7, 10, 10}
 	for i, dur := range durations {
-		require.Equal(t, time.Millisecond*time.Duration(dur*1000), strategy.Duration(i))
+		require.Equal(t, time.Millisecond*dur*1000, strategy.Duration(i))
```
Comment (protolambda) [inline op-service/backoff/strategies_test.go]: Would revert this, or change the values in the array to multiples of `time.Millisecond`

**Review by protolambda — CHANGES_REQUESTED** [review]

File: op-service/backoff/strategies.go (hunk 5 above)
Comment (Rcontre360) [reply op-service/backoff/strategies.go]: Got it

File: op-service/backoff/strategies.go [hunk 7]
```diff
@@ -16,24 +16,23 @@ type Strategy interface {
...
+	// MaxJitter is the maximum amount of random jitter to insert between attempts.
+	MaxJitter time.Duration
 }
 
 func (e *ExponentialStrategy) Duration(attempt int) time.Duration {
-	var jitter int
+	var jitter time.Duration
 	if e.MaxJitter > 0 {
-		jitter = rand.Intn(e.MaxJitter)
+		jitter = time.Duration(rand.Int63n(e.MaxJitter.Milliseconds()))
```
Comment (trianglesphere) [inline op-service/backoff/strategies.go]: `time.Duration` is in nanoseconds, when creating one, you always need to multiple by the unit.

**Review by Rcontre360 — COMMENTED** [review]
There's only one place where ExponentialStrategy is used, it is /op-node/sources/sync_client.go. It is assumed in the expStrategy. Duration function that the fields in the expStrategy will be provided correctly (example: Max: time.Duration(1000 * time.Millisecond)).

File: op-service/backoff/strategies.go [hunk 8]
```diff
@@ -29,15 +29,15 @@ type ExponentialStrategy struct {
...
-	dur := e.Min + time.Duration((math.Pow(2, float64(attempt)) * 1000))
+	dur := e.Min + time.Duration(int(math.Pow(2, float64(attempt))*1000))*time.Millisecond
 	dur += jitter
 	if dur > e.Max {
-		return time.Millisecond * time.Duration(e.Max)
+		return e.Max
 	}
 
-	return time.Millisecond * time.Duration(dur)
+	return time.Duration(dur)
```
Comment (protolambda) [inline op-service/backoff/strategies.go]:
```suggestion
	return dur
```
It's already a `time.Duration`

File: op-service/backoff/strategies_test.go [hunk 9]
```diff
@@ -9,13 +9,14 @@ import (
...
+		Min:       3000 * time.Millisecond,
+		Max:       10000 * time.Millisecond,
 		MaxJitter: 0,
 	}
 
 	durations := []time.Duration{4, 5, 7, 10, 10}
 	for i, dur := range durations {
-		require.Equal(t, time.Millisecond*dur*1000, strategy.Duration(i))
+		msDuration := dur * time.Millisecond * 1000
+		require.Equal(t, msDuration, strategy.Duration(i))
```
Comment (protolambda) [inline op-service/backoff/strategies_test.go]:
```suggestion
		require.Equal(t, dur*time.Second, strategy.Duration(i))
```

File: op-service/backoff/strategies.go (hunk 7 above)
Comment (Rcontre360) [reply op-service/backoff/strategies.go]: Yes, thanks! didnt know it when making the changes

**Review by protolambda — COMMENTED** [review]
Two more nits, then done

**Comment by protolambda** [comment]
And `ExponentialStrategy` in `sync_client.go` needs the `* time.Millisecond` to have the proper units

**Comment by Rcontre360** [comment]
Yes @protolambda I already changed op-node/sources/sync_client.go to have the correct units :+1:

**Review by protolambda — APPROVED** [review]

**Comment by tynes** [comment]
@Rcontre360 if you rebase, CI should be passing again

**Comment by Rcontre360** [comment]
> @Rcontre360 if you rebase, CI should be passing again

hmmm. Just rebased and they keep failing :thinking: . Is this happening to other PRs as well?

**Comment by protolambda** [comment]
@Rcontre360 likely other PRs as well, we're looking into it

**Comment by protolambda** [comment]
@Mergifyio squash title+body

**Comment by protolambda** [comment]
@Mergifyio squash title+body

**Comment by protolambda** [comment]
Ugh, mergify doesn't preserve the co-author info, let me fix that

**Comment by protolambda** [comment]
fixed committer info, and rebased on develop to fix the flake

**Pull request merged**